pkg/operator/controllers/machineset/machineset_controller_test.go

```diff
@@ -0,0 +1,415 @@
+		},
+	}
+
+	tests := []struct {
+		name           string
+		maodata        []runtime.Object
+		mocks          func(maocli *maofake.Clientset)
+		wantReplicas   int32
+		assertReplicas bool
+		wantErr        string
+		objectName     string
+	}{
+		{
+			name:           "no worker replicas, machineset-0 modified",
+			maodata:        FakeMachineSets(0, 0, 0),
+			wantReplicas:   2, // expected replica count after reconcile
+			assertReplicas: true,
+			wantErr:        "",
+			objectName:     "aro-fake-machineset-0",
+		},
+		{
+			name:           "one worker replica, machineset-0 modified",
+			maodata:        FakeMachineSets(1, 0, 0),
+			wantReplicas:   2,
+			assertReplicas: true,
+			wantErr:        "",
+			objectName:     "aro-fake-machineset-0",
+		},
+		{
+			name:           "two worker replicas, machineset-0 modified",
+			maodata:        FakeMachineSets(1, 1, 0),
+			wantReplicas:   1,
+			assertReplicas: true,
+			wantErr:        "",
+			objectName:     "aro-fake-machineset-0",
+		},
+		{
+			name:           "three worker replicas, machineset-0 modified",
+			maodata:        FakeMachineSets(1, 1, 1),
+			wantReplicas:   1,
+			assertReplicas: true,
+			wantErr:        "",
+			objectName:     "aro-fake-machineset-0",
+		},
+		{
+			name:           "three worker replicas in machineset-1, machineset-0 modified",
+			maodata:        FakeMachineSets(0, 3, 0),
+			wantReplicas:   0,
+			assertReplicas: true,
+			wantErr:        "",
+			objectName:     "aro-fake-machineset-0",
+		},
+		{
+			name:           "two worker replicas in machineset-1, machineset-0 modified",
+			maodata:        FakeMachineSets(0, 2, 0),
+			wantReplicas:   0,
+			assertReplicas: true,
+			wantErr:        "",
+			objectName:     "aro-fake-machineset-0",
+		},
+		{
+			name:    "machineset-0 not found",
+			maodata: FakeMachineSets(2, 0, 0),
+			mocks: func(maocli *maofake.Clientset) {
+				maocli.PrependReactor("get", "machinesets", func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, nil, &kerrors.StatusError{ErrStatus: metav1.Status{
+						Message: "machineset-0 not found",
+						Reason:  metav1.StatusReasonNotFound,
+					}}
+				})
+			},
+			assertReplicas: false,
+			wantErr:        "machineset-0 not found",
+			objectName:     "aro-fake-machineset-0",
+		},
+		{
+			name:    "machineset-0 not found, object is no longer present",
+			maodata: FakeMachineSets(1, 0, 0),
+			mocks: func(maocli *maofake.Clientset) {
+				maocli.PrependReactor("get", "machinesets", func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
+					return true, nil, &kerrors.StatusError{ErrStatus: metav1.Status{
+						Message: "machineset-0 not found, object is no longer present",
+						Reason:  metav1.StatusReasonGone,
+					}}
+				})
+			},
+			assertReplicas: false,
+			wantErr:        "machineset-0 not found, object is no longer present",
+			objectName:     "aro-fake-machineset-0",
+		},
```

Comment:
Almost, but it uses a different error type so that we get coverage for the `return reconcile.Result{}, err` statement.

Reply:
Right... That makes sense. But I would raplace this one with any random error. Something simple like `errors.New("fake error")` instead of using `kerrors.StatusError` because we only care about status error with reason `metav1.StatusReasonNotFound` and "everything else".

Reply:
Sure, that works too :+1: